pkg/blobaccess/file/access.go

```diff
@@ -0,0 +1,338 @@
+	ok, err := vfs.Exists(a.fs, a.path)
+	if err != nil {
+		return err
+	}
+	if !ok {
+		return errors.ErrNotFound("file", a.path)
+	}
+	return nil
+}
+
+func (a *fileDataAccess) Origin() string {
+	return a.path
+}
+
+////////////////////////////////////////////////////////////////////////////////
+
+type fileBlobAccess struct {
+	fileDataAccess
+	mimeType string
+}
+
+var (
+	_ bpi.BlobAccess   = (*fileBlobAccess)(nil)
+	_ bpi.FileLocation = (*fileBlobAccess)(nil)
+)
+
+func (f *fileBlobAccess) FileSystem() vfs.FileSystem {
+	return f.fs
+}
+
+func (f *fileBlobAccess) Path() string {
+	return f.path
+}
+
+func (f *fileBlobAccess) Dup() (bpi.BlobAccess, error) {
+	return f, nil
+}
+
+func (f *fileBlobAccess) Size() int64 {
+	size := bpi.BLOB_UNKNOWN_SIZE
+	fi, err := f.fs.Stat(f.path)
+	if err == nil {
+		size = fi.Size()
+	}
+	return size
+}
+
+func (f *fileBlobAccess) MimeType() string {
+	return f.mimeType
+}
+
+func (f *fileBlobAccess) DigestKnown() bool {
+	return false
+}
+
+func (f *fileBlobAccess) Digest() digest.Digest {
+	r, err := f.Reader()
+	if err != nil {
+		return ""
+	}
+	defer r.Close()
+	d, err := digest.FromReader(r)
+	if err != nil {
+		return ""
+	}
+	return d
+}
+
+// BlobAccess wraps a file path into a BlobAccess, which does not need a close.
+func BlobAccess(mime string, path string, fss ...vfs.FileSystem) bpi.BlobAccess {
+	return &fileBlobAccess{
+		mimeType:       mime,
+		fileDataAccess: fileDataAccess{fs: utils.FileSystem(fss...), path: path},
+	}
+}
+
+func ProviderForFile(mime string, path string, fss ...vfs.FileSystem) bpi.BlobAccessProvider {
+	return bpi.BlobAccessProviderFunction(func() (bpi.BlobAccess, error) {
+		return BlobAccess(mime, path, fss...), nil
+	})
+}
+
+type fileBlobAccessView struct {
+	_blobAccess
+	access *fileDataAccess
+}
+
+var (
+	_ bpi.BlobAccess   = (*fileBlobAccessView)(nil)
+	_ bpi.FileLocation = (*fileBlobAccessView)(nil)
+)
+
+func (f *fileBlobAccessView) Dup() (bpi.BlobAccess, error) {
+	b, err := f._blobAccess.Dup()
+	if err != nil {
+		return nil, err
+	}
+	return &fileBlobAccessView{b, f.access}, nil
+}
+
+func (f *fileBlobAccessView) FileSystem() vfs.FileSystem {
+	return f.access.fs
+}
+
+func (f *fileBlobAccessView) Path() string {
+	return f.access.path
+}
+
+func BlobAccessWithCloser(closer io.Closer, mime string, path string, fss ...vfs.FileSystem) bpi.BlobAccess {
+	fb := &fileBlobAccess{fileDataAccess{fs: utils.FileSystem(fss...), path: path}, mime}
+	return &fileBlobAccessView{
+		bpi.NewBlobAccessForBase(fb, closer),
+		&fb.fileDataAccess,
+	}
+}
+
+////////////////////////////////////////////////////////////////////////////////
+
+type temporaryFileBlob struct {
+	_blobAccess
+	lock       sync.Mutex
+	path       string
+	file       vfs.File
+	filesystem vfs.FileSystem
+}
+
+var (
+	_ bpi.BlobAccessBase = (*temporaryFileBlob)(nil)
+	_ bpi.FileLocation   = (*temporaryFileBlob)(nil)
+)
+
+func (b *temporaryFileBlob) Validate() error {
```

Comment:
Documentation? This `Validate()` seem to work differently, since it uses locks...

Reply:
it also validates, but to do so, it has to request a lock first, here.